Sources/StreamChat/Database/DTOs/ChannelDTO.swift

```diff
@@ -436,70 +436,76 @@ extension ChatChannel {
-                ),
-                NSPredicate(format: "createdAt > %@", currentUserRead?.lastReadAt.bridgeDate ?? DBDate(timeIntervalSince1970: 0)),
-                NSPredicate(format: "%@ IN mentionedUsers", currentUser.user)
-            ])
-
-            do {
-                return ChannelUnreadCount(
-                    messages: allUnreadMessages,
-                    mentions: try context.count(for: unreadMentionsRequest)
-                )
-            } catch {
-                log.error("Failed to fetch unread counts for channel `\(cid)`. Error: \(error)")
+            // Therefore, no unread messages with mentions and we can skip the fetch
+            if allUnreadMessages == 0 {
                 return .noUnread
             }
+            let unreadMentionsCount = dto.messages
+                .sorted(by: { $0.createdAt.bridgeDate > $1.createdAt.bridgeDate })
+                .prefix(allUnreadMessages)
+                .filter { $0.mentionedUsers.contains(currentUserDTO.user) }
+                .count
+            return ChannelUnreadCount(
+                messages: allUnreadMessages,
+                mentions: unreadMentionsCount
+            )
         }()
 
         let messages: [ChatMessage] = {
-            MessageDTO
-                .load(
-                    for: dto.cid,
-                    limit: dto.managedObjectContext?.localCachingSettings?.chatChannel.latestMessagesLimit ?? 25,
-                    deletedMessagesVisibility: dto.managedObjectContext?.deletedMessagesVisibility ?? .visibleForCurrentUser,
-                    shouldShowShadowedMessages: dto.managedObjectContext?.shouldShowShadowedMessages ?? false,
-                    context: context
-                )
+            guard !dto.messages.isEmpty else { return [] }
+            let request = MessageDTO.channelMessagesRequest(
+                for: dto.cid,
+                limit: dto.managedObjectContext?.localCachingSettings?.chatChannel.latestMessagesLimit ?? 25,
+                deletedMessagesVisibility: dto.managedObjectContext?.deletedMessagesVisibility ?? .visibleForCurrentUser,
+                shouldShowShadowedMessages: dto.managedObjectContext?.shouldShowShadowedMessages ?? false
+            )
+            return dto.messages
+                .filtered(using: request)
                 .compactMap { try? $0.relationshipAsModel(depth: depth) }
         }()
 
         let latestMessageFromUser: ChatMessage? = {
-            guard let currentUser = context.currentUser else { return nil }
-
-            return try? MessageDTO
-                .loadLastMessage(
-                    from: currentUser.user.id,
-                    in: dto.cid,
-                    context: context
-                )?
+            guard let currentUserId = context.currentUser?.user.id else { return nil }
+            return try? dto.messages
+                .sorted(by: { $0.createdAt.bridgeDate > $1.createdAt.bridgeDate })
```

Comment:
We are doing this sorting in multiple places, we probably can just do it once and reuse it. Something like:
`let latestMessages = dto.messages.sorted(by: { $0.createdAt.bridgeDate > $1.createdAt.bridgeDate })`

And then use `latestMessages` everywhere